### PR TITLE
Collect cluster state into a single type.

### PR DESCRIPTION
**Pull request by mml**
Mostly just makes it easier to read and, hopefully, extend.

### CODE REVIEW

**Comment by mml** [comment]
@madhusudancs

File: test/e2e/federated-service.go [hunk 1]
```diff
@@ -62,11 +62,22 @@ var FederatedServiceLabels = map[string]string{
 	"foo": "bar",
 }
 
+/*
+type cluster keeps track of the assorted objects and state related to each
```
Comment (madhusudancs) [inline test/e2e/federated-service.go]: Convention is to just start with the entity's name - "cluster keeps track of ..."

File: test/e2e/federated-service.go [hunk 2]
```diff
@@ -106,58 +117,60 @@ var _ = framework.KubeDescribe("[Feature:Federation]", func() {
 			}
 			framework.Logf("%d clusters are Ready", len(contexts))
 
-			clusterClientSets = make([]*release_1_3.Clientset, len(clusterList.Items))
-			for i, cluster := range clusterList.Items {
-				framework.Logf("Creating a clientset for the cluster %s", cluster.Name)
+			// clusters = make([]cluster, len(clusterList.Items))
+			clusters = map[string]*cluster{}
+			primaryClusterName = clusterList.Items[0].Name
```
Comment (madhusudancs) [inline test/e2e/federated-service.go]: We use the last cluster as the primary cluster in our setup/teardown scripts - https://github.com/kubernetes/kubernetes/blob/master/hack/e2e-internal/e2e-up.sh#L34

File: test/e2e/federated-service.go [hunk 3]
```diff
@@ -501,36 +497,42 @@ func createBackendPodsOrFail(clusterClientSets []*release_1_3.Clientset, namespa
...
+		By(fmt.Sprintf("Successfully created pod %q in namespace %q in cluster %q: %v", pod.Name, namespace, name, *createdPod))
+		c.backendPod = createdPod
 	}
-	return pods
 }
 
 /*
-deleteBackendPodsOrFail deletes one pod from each cluster (unless pods[n] is nil for that cluster)
-If deletion of any pod fails, the test fails  (possibly with a partially deleted set of pods). No retries are attempted.
+deletes exactly one backend pod which must not be nil
```
Comment (madhusudancs) [inline test/e2e/federated-service.go]: "deleteOneBackendPodOrFail deletes ..."

**Comment by madhusudancs** [comment]
A few minor nits, LGTM otherwise. Feel free to apply the label after addressing the comments.

File: test/e2e/federated-service.go [hunk 4]
```diff
@@ -62,11 +62,22 @@ var FederatedServiceLabels = map[string]string{
...
+	*release_1_3.Clientset
+	namespaceCreated bool    // Did we need to create a new namespace in this cluster?  If so, we should delete it.
+	backendPod       *v1.Pod // The backend pod, if one's been created.
+}
+
 var _ = framework.KubeDescribe("[Feature:Federation]", func() {
 	f := framework.NewDefaultFederatedFramework("federated-service")
-	var clusterClientSets []*release_1_3.Clientset
-	var clusterNamespaceCreated []bool // Did we need to create a new namespace in each of the above clusters?  If so, we should delete it.
+	var clusters map[string]*cluster
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: nit: To be pedantic, perhaps mention that the key is the cluster name?

File: test/e2e/federated-service.go [hunk 5]
```diff
@@ -106,58 +117,60 @@ var _ = framework.KubeDescribe("[Feature:Federation]", func() {
 			}
 			framework.Logf("%d clusters are Ready", len(contexts))
 
-			clusterClientSets = make([]*release_1_3.Clientset, len(clusterList.Items))
-			for i, cluster := range clusterList.Items {
-				framework.Logf("Creating a clientset for the cluster %s", cluster.Name)
+			// clusters = make([]cluster, len(clusterList.Items))
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: nit: Remove unused code.

File: test/e2e/federated-service.go (hunk 2 above)
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: Also, the ordering of clusterlist depends on the ordering of the list returned by the API call to list clusters, which is not, as far as I know, well defined.  I think that we need a more definitive means of determining the "primary cluster".  It's the cluster in which the federation control plane is running (irrespective of where it appears in the list).

File: test/e2e/federated-service.go [hunk 6]
```diff
@@ -106,58 +117,60 @@ var _ = framework.KubeDescribe("[Feature:Federation]", func() {
...
+			clusters = map[string]*cluster{}
+			primaryClusterName = clusterList.Items[0].Name
+			By(fmt.Sprintf("Labeling %q as the first cluster", primaryClusterName))
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: nit: You're not actually labelling it, are you?

File: test/e2e/federated-service.go [hunk 7]
```diff
@@ -106,58 +117,60 @@ var _ = framework.KubeDescribe("[Feature:Federation]", func() {
...
 				// The e2e Framework created the required namespace in one of the clusters, but we need to create it in all the others, if it doesn't yet exist.
-				if _, err := cs.Core().Namespaces().Get(f.Namespace.Name); errors.IsNotFound(err) {
+				if _, err := c.Clientset.Core().Namespaces().Get(f.Namespace.Name); errors.IsNotFound(err) {
 					ns := &v1.Namespace{
 						ObjectMeta: v1.ObjectMeta{
 							Name: f.Namespace.Name,
 						},
 					}
-					_, err := cs.Core().Namespaces().Create(ns)
+					_, err := c.Clientset.Core().Namespaces().Create(ns)
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: nit: I would prefer us to use f.Namespace.Name here, in preference to 'ns'. I'm not even sure where the latter is declared - I suspect that it's redundant, and should be removed from the framework.  At the very least. the variable used here to create the namespace, and the one used below to report success or failure, should be the same variable.

Comment (quinton-hoole) [inline test/e2e/federated-service.go]: Oh, forget it, I'm completely confused.

**Comment by quinton-hoole** [comment]
LGTM barring nits.  @mml as discussed, lets get the existing e2e tests consistently passing before merging this PR. I think we're just one PR away from the former.

File: test/e2e/federated-service.go (hunk 6 above)
Comment (mml) [inline test/e2e/federated-service.go]: Marking?  Recording?

**Comment by quinton-hoole** [comment]
I think you want 5b8113d in a separate PR?

**Comment by mml** [comment]
@quinton-hoole that commit addresses comments from this review.  It'll get squashed at the end.

**Comment by quinton-hoole** [comment]
LGTM added now that Jenkins tests are all green. 
This PR probably needs a squash before merging.
Feel free to re-add the LGTM once rebase and squash are complete.

**Comment by mml** [comment]
Too late.  No squash!

**Comment by mml** [comment]
Oops, this PR also merged's @madhusudancs's https://github.com/kubernetes/kubernetes/pull/28388/commits/d618e5329bfb86c266ab324add9bcf10cdb77bbc.

It would be nice to be able to add a do-not-merge label that could prevent accidental merging in this case.

**Comment by quinton-hoole** [comment]
@mml  There is a "do-not-merge" label that does what you want.  It's not clear to my why this did not cause a merge conflict and therefore require a rebase.  Could you open an issue to understand and if necessary fix that.  @lavalamp FYI.  This is the third time I've seen such a mysterious lack of merge conflicts.

**Comment by mml** [comment]
The lack of merge conflicts isn't surprising to me.  My PR included exactly madhu's original commit, which had merged just before, so in the end that part of my merge was a nop.

But I now have do-not-merge in my toolbox.  Thanks.

**Comment by madhusudancs** [comment]
Yeah, lack of merge conflict is the expected behavior since @mml's version and my version matched.

**Pull request merged**